In [14]:
#!pip install --upgrade tensorflow
#!pip install --upgrade tensorflow-io

  Using cached tensorflow-2.14.0-cp311-cp311-win_amd64.whl (2.1 kB)
     ---------------------------------------- 0.0/284.2 MB ? eta -:--:--
     ---------------------------------------- 0.1/284.2 MB 1.7 MB/s eta 0:02:53
     ---------------------------------------- 0.2/284.2 MB 2.1 MB/s eta 0:02:16
     ---------------------------------------- 0.3/284.2 MB 2.1 MB/s eta 0:02:14
     ---------------------------------------- 0.4/284.2 MB 2.6 MB/s eta 0:01:51
     ---------------------------------------- 0.6/284.2 MB 2.9 MB/s eta 0:01:38
     ---------------------------------------- 0.6/284.2 MB 2.9 MB/s eta 0:01:38
     ---------------------------------------- 0.6/284.2 MB 2.2 MB/s eta 0:02:08
     ---------------------------------------- 0.6/284.2 MB 2.2 MB/s eta 0:02:08
     ---------------------------------------- 0.6/284.2 MB 2.2 MB/s eta 0:02:08
     ---------------------------------------- 0.6/284.2 MB 2.2 MB/s eta 0:02:08
     ---------------------------------------- 0.6/284.2 MB 

In [15]:
#!pip install spacy

In [1]:
import numpy as np
import pandas as pd
import spacy
import re

In [2]:
data=pd.read_csv(r"D:/study meterials/datasets/cnn_dailymail/train.csv")
test=pd.read_csv(r"D:/study meterials/datasets/cnn_dailymail/test.csv")
validation=pd.read_csv(r"D:/study meterials/datasets/cnn_dailymail/validation.csv")

In [3]:
data.head()

id   
0  0001d1afc246a7964130f43ae940af6bc6c57f01  \
1  0002095e55fcbd3a2f366d9bf92a95433dc305ef   
2  00027e965c8264c35cc1bc55556db388da82b07f   
3  0002c17436637c4fe1837c935c04de47adb18e9a   
4  0003ad6ef0c37534f80b55b4235108024b407f0b   

                                             article   
0  By . Associated Press . PUBLISHED: . 14:11 EST...  \
1  (CNN) -- Ralph Mata was an internal affairs li...   
2  A drunk driver who killed a young woman in a h...   
3  (CNN) -- With a breezy sweep of his pen Presid...   
4  Fleetwood are the only team still to have a 10...   

                                          highlights  
0  Bishop John Folda, of North Dakota, is taking ...  
1  Criminal complaint: Cop used his role to help ...  
2  Craig Eccleston-Todd, 27, had drunk at least t...  
3  Nina dos Santos says Europe must be ready to a...  
4  Fleetwood top of League One after 2-0 win at S...

In [4]:
validation.head()

id   
0  61df4979ac5fcc2b71be46ed6fe5a46ce7f071c3  \
1  21c0bd69b7e7df285c3d1b1cf56d4da925980a68   
2  56f340189cd128194b2e7cb8c26bb900e3a848b4   
3  00a665151b89a53e5a08a389df8334f4106494c2   
4  9f6fbd3c497c4d28879bebebea220884f03eb41a   

                                             article   
0  Sally Forrest, an actress-dancer who graced th...  \
1  A middle-school teacher in China has inked hun...   
2  A man convicted of killing the father and sist...   
3  Avid rugby fan Prince Harry could barely watch...   
4  A Triple M Radio producer has been inundated w...   

                                          highlights  
0  Sally Forrest, an actress-dancer who graced th...  
1  Works include pictures of Presidential Palace ...  
2  Iftekhar Murtaza, 29, was convicted a year ago...  
3  Prince Harry in attendance for England's crunc...  
4  Nick Slater's colleagues uploaded a picture to...

In [5]:
test.head()

id   
0  92c514c913c0bdfe25341af9fd72b29db544099b  \
1  2003841c7dc0e7c5b1a248f9cd536d727f27a45a   
2  91b7d2311527f5c2b63a65ca98d21d9c92485149   
3  caabf9cbdf96eb1410295a673e953d304391bfbb   
4  3da746a7d9afcaa659088c8366ef6347fe6b53ea   

                                             article   
0  Ever noticed how plane seats appear to be gett...  \
1  A drunk teenage boy had to be rescued by secur...   
2  Dougie Freedman is on the verge of agreeing a ...   
3  Liverpool target Neto is also wanted by PSG an...   
4  Bruce Jenner will break his silence in a two-h...   

                                          highlights  
0  Experts question if  packed out planes are put...  
1  Drunk teenage boy climbed into lion enclosure ...  
2  Nottingham Forest are close to extending Dougi...  
3  Fiorentina goalkeeper Neto has been linked wit...  
4  Tell-all interview with the reality TV star, 6...

In [6]:
print(test.loc[:1, 'article'].tolist())

["Ever noticed how plane seats appear to be getting smaller and smaller? With increasing numbers of people taking to the skies, some experts are questioning if having such packed out planes is putting passengers at risk. They say that the shrinking space on aeroplanes is not only uncomfortable - it's putting our health and safety in danger. More than squabbling over the arm rest, shrinking space on planes putting our health and safety in danger? This week, a U.S consumer advisory group set up by the Department of Transportation said at a public hearing that while the government is happy to set standards for animals flying on planes, it doesn't stipulate a minimum amount of space for humans. 'In a world where animals have more rights to space and food than humans,' said Charlie Leocha, consumer representative on the committee.\xa0'It is time that the DOT and FAA take a stand for humane treatment of passengers.' But could crowding on planes lead to more serious issues than fighting for s

In [7]:
print(test.loc[:1, 'highlights'].tolist())

['Experts question if  packed out planes are putting passengers at risk .\nU.S consumer advisory group says minimum space must be stipulated .\nSafety tests conducted on planes with more leg room than airlines offer .', "Drunk teenage boy climbed into lion enclosure at zoo in west India .\nRahul Kumar, 17, ran towards animals shouting 'Today I kill a lion!'\nFortunately he fell into a moat before reaching lions and was rescued ."]


In [8]:
data.columns

Index(['id', 'article', 'highlights'], dtype='object')

In [9]:
df = data
df = df.highlights.dropna()
df_articles = df.values.tolist()

# Limit the length of df_articles to 1000
df_articles = df_articles[:1000]

len(df_articles)

1000

In [27]:
df_articles[0:4]

['Bishop John Folda, of North Dakota, is taking time off after being diagnosed .\nHe contracted the infection through contaminated food in Italy .\nChurch members in Fargo, Grand Forks and Jamestown could have been exposed .',
 'Criminal complaint: Cop used his role to help cocaine traffickers .\nRalph Mata, an internal affairs lieutenant, allegedly helped group get guns .\nHe also arranged to pay two assassins in a murder plot, a complaint alleges .',
 "Craig Eccleston-Todd, 27, had drunk at least three pints before driving car .\nWas using phone when he veered across road in Yarmouth, Isle of Wight .\nCrashed head-on into 28-year-old Rachel Titley's car, who died in hospital .\nPolice say he would have been over legal drink-drive limit at time of crash .\nHe was found guilty at Portsmouth Crown Court of causing death by dangerous driving .",
 "Nina dos Santos says Europe must be ready to accept sanctions will hurt both sides .\nTargeting Russia's business community would be one way o

In [25]:
#!pip install -U transformers
#!pip install -U torch

<h2>Summarization using T5 model

In [10]:
import torch
import os
import json

from transformers import T5Tokenizer, T5ForConditionalGeneration, T5Config

In [28]:
#!pip install sentencepiece

     ---------------------------------------- 0.0/977.5 kB ? eta -:--:--
     ---------------------------------------- 10.2/977.5 kB ? eta -:--:--
     -- ---------------------------------- 61.4/977.5 kB 656.4 kB/s eta 0:00:02
     -- ---------------------------------- 61.4/977.5 kB 656.4 kB/s eta 0:00:02
     --- --------------------------------- 92.2/977.5 kB 476.3 kB/s eta 0:00:02
     --- --------------------------------- 92.2/977.5 kB 476.3 kB/s eta 0:00:02
     ---- ------------------------------- 112.6/977.5 kB 437.6 kB/s eta 0:00:02
     ---- ------------------------------- 122.9/977.5 kB 400.9 kB/s eta 0:00:03
     ----- ------------------------------ 143.4/977.5 kB 405.9 kB/s eta 0:00:03
     ------ ----------------------------- 163.8/977.5 kB 364.0 kB/s eta 0:00:03
     ------ ----------------------------- 174.1/977.5 kB 361.7 kB/s eta 0:00:03
     ------ ----------------------------- 174.1/977.5 kB 361.7 kB/s eta 0:00:03
     ------- ---------------------------- 194.6/977.5

In [11]:
T5_PATH = 'Einmalumdiewelt/T5-Base_GNAD'
t5_model = T5ForConditionalGeneration.from_pretrained(T5_PATH, output_past=True)
t5_tokenizer = T5Tokenizer.from_pretrained(T5_PATH)
x = t5_tokenizer(df_articles, truncation=True, max_length=1000)

C:\Users\ASUS\anaconda3\envs\notebook-6.5.4\Lib\site-packages\huggingface_hub\file_download.py:137: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\ASUS\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


You are using the default legacy behaviour of the <class 'transformers.models.t5.tokenization_t5.T5Tokenizer'>. This is expected, and simply means that the `legacy` (previous) behavior will be used so nothing changes for you. If you want to use the new behaviour, set `legacy=False`. This should only be set if you understand what it means, and thouroughly read the reason why this was added as explained in https://github.com/huggingface/transformers/pull/24565
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


In [12]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

def t5_summarize(input_text, num_beams=4, num_words=80):
    input_text = ' '.join(input_text.split())
    input_tokenized = t5_tokenizer.encode(input_text, return_tensors="pt").to(device)
    summary_task = torch.tensor([[21603, 10]]).to(device)
    input_tokenized = torch.cat([summary_task, input_tokenized], dim=-1).to(device)

    # Move the model to the same device as input tensors
    t5_model.to(device)

    summary_ids = t5_model.generate(input_tokenized,
                                    num_beams=int(num_beams),
                                    no_repeat_ngram_size=3,
                                    length_penalty=2.0,
                                    min_length=30,
                                    max_length=int(num_words),
                                    early_stopping=True)
    output = [t5_tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=False) 
              for g in summary_ids]
    return output[0] 

In [13]:
%%time
for i in range(4):
    
        print('news articles ',i + 1, " : \n" )
        print(t5_summarize(df_articles[i]))
        print('............................................................................\n\n\n\n')

news articles  1  : 

In North Dakota, a priest has been diagnosed with a food-borne illness. He is taking time off after being diagnosed with the infection.
............................................................................




news articles  2  : 

Cop allegedly arranged to pay two assassins in a murder plot. He also allegedly helped group get guns, a criminal complaint alleges.
............................................................................




news articles  3  : 

Craig Eccleston-Todd, 27, was using phone when he veered across road in Yarmouth, Isle of Wight . He had drunk at least three pints before driving car .
............................................................................




news articles  4  : 

In a recent interview, the US-based economist Nina dos Santos has said that targeting Russia's business community would be one way of sapping their support for President Putin.
......................................................................

In [29]:
data['article'][0]

"By . Associated Press . PUBLISHED: . 14:11 EST, 25 October 2013 . | . UPDATED: . 15:36 EST, 25 October 2013 . The bishop of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A virus in late September and early October. The state Health Department has issued an advisory of exposure for anyone who attended five churches and took communion. Bishop John Folda (pictured) of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A . State Immunization Program Manager Molly Howell says the risk is low, but officials feel it's important to alert people to the possible exposure. The diocese announced on Monday that Bishop John Folda is taking time off after being diagnosed with hepatitis A. The diocese says he contracted the infection through contaminated food while attending a conference for newly ordained 

In [31]:
data['highlights'][0]

'Bishop John Folda, of North Dakota, is taking time off after being diagnosed .\nHe contracted the infection through contaminated food in Italy .\nChurch members in Fargo, Grand Forks and Jamestown could have been exposed .'

In [30]:
def summarize_text(text):
    # Generate summary using the T5 model
    summary = t5_summarize(text)
    
    # Show the generated summary
    print("\nSummary:")
    print(summary)

# Prompt the user for input and generate summary
user_input = input("Enter the text to summarize: ")
summarize_text(user_input)

Enter the text to summarize: "By . Associated Press . PUBLISHED: . 14:11 EST, 25 October 2013 . | . UPDATED: . 15:36 EST, 25 October 2013 . The bishop of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A virus in late September and early October. The state Health Department has issued an advisory of exposure for anyone who attended five churches and took communion. Bishop John Folda (pictured) of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A . State Immunization Program Manager Molly Howell says the risk is low, but officials feel it's important to alert people to the possible exposure. The diocese announced on Monday that Bishop John Folda is taking time off after being diagnosed with hepatitis A. The diocese says he contracted the infection through contaminated food while attending a c

<h2>Summarization using bart model

In [15]:
from transformers import BartTokenizer, BartForConditionalGeneration, BartConfig

In [16]:
bart_PATH = 'facebook/bart-large-cnn'
bart_model = BartForConditionalGeneration.from_pretrained(bart_PATH, output_past=True)
bart_tokenizer = BartTokenizer.from_pretrained(bart_PATH)
x = bart_tokenizer(df_articles, truncation=True, max_length=500)

In [17]:
def bart_summarize(input_text, num_beams=4, num_words=80):
    #input_text = str(input_text)
    input_text = ' '.join(input_text.split())
    input_tokenized = bart_tokenizer.encode(input_text, return_tensors='pt')
    summary_ids = bart_model.generate(input_tokenized,
                                      num_beams=int(num_beams),
                                      no_repeat_ngram_size=3,
                                      length_penalty=2.0,
                                      min_length=80,
                                      max_length=int(num_words),
                                      early_stopping=True)
    output = [bart_tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=False) for g in summary_ids]
    return output[0]

In [18]:
%%time
for i in range(4):
    
        print('news article ',i + 1, " : \n" )
        print(bart_summarize(df_articles[i]))
        print('............................................................................\n\n\n\n')

news article  1  : 

Bishop John Folda, of North Dakota, is taking time off after being diagnosed. He contracted the infection through contaminated food in Italy. Church members in Fargo, Grand Forks and Jamestown could have been exposed. The bishop is expected to return to work in the fall. He has been diagnosed with an infection that can be transmitted via contaminated food. The infection can also be
............................................................................




news article  2  : 

Criminal complaint: Cop used his role to help cocaine traffickers. Ralph Mata, an internal affairs lieutenant, allegedly helped group get guns. He also arranged to pay two assassins in a murder plot, a complaint alleges. Mata is accused of using his position to help drug traffickers get guns and money. He is also accused of arranging to pay assassins to kill a man in a plot.
............................................................................




news article  3  : 

Craig Eccles

In [77]:
data['article'][4]

"Fleetwood are the only team still to have a 100% record in Sky Bet League One as a 2-0 win over Scunthorpe sent Graham Alexander’s men top of the table. The Cod Army are playing in the third tier for the first time in their history after six promotions in nine years and their remarkable ascent shows no sign of slowing with Jamie Proctor and Gareth Evans scoring the goals at Glanford Park. Fleetwood were one of five teams to have won two out of two but the other four clubs - Peterborough, Bristol City, Chesterfield and Crawley - all hit their first stumbling blocks. Posh were defeated 2-1 by Sheffield United, who had lost both of their opening contests. Jose Baxter’s opener gave the Blades a first-half lead, and although it was later cancelled out by Shaun Brisley’s goal, Ben Davies snatched a winner six minutes from time. In the lead: Jose Baxter (right) celebrates opening the scoring for Sheffield United . Up for the battle: Sheffield United's Michael Doyle (left) challenges Peterbor

In [78]:
data['highlights'][4]

'Fleetwood top of League One after 2-0 win at Scunthorpe .\nPeterborough, Bristol City, Chesterfield and Crawley all drop first points of the season .\nStand-in striker Matt Done scores a hat-trick as Rochdale thrash Crewe 5-2 .\nWins for Notts County and Yeovil .\nCoventry/Bradford and Oldham/Port Vale both end in draws .\nA late Stephen Bywater own goal denies Gillingham three points against Millwall .'

In [79]:
def summarize_text(text):
    # Generate summary using the bart model
    summary = bart_summarize(text)
    
    # Show the generated summary
    print("\nSummary:")
    print(summary)

# Prompt the user for input and generate summary
user_input = input("Enter the text to summarize: ")
summarize_text(user_input)

Enter the text to summarize: Fleetwood are the only team still to have a 100% record in Sky Bet League One as a 2-0 win over Scunthorpe sent Graham Alexander’s men top of the table. The Cod Army are playing in the third tier for the first time in their history after six promotions in nine years and their remarkable ascent shows no sign of slowing with Jamie Proctor and Gareth Evans scoring the goals at Glanford Park. Fleetwood were one of five teams to have won two out of two but the other four clubs - Peterborough, Bristol City, Chesterfield and Crawley - all hit their first stumbling blocks. Posh were defeated 2-1 by Sheffield United, who had lost both of their opening contests. Jose Baxter’s opener gave the Blades a first-half lead, and although it was later cancelled out by Shaun Brisley’s goal, Ben Davies snatched a winner six minutes from time. In the lead: Jose Baxter (right) celebrates opening the scoring for Sheffield United . Up for the battle: Sheffield United's Michael Doyl

In [86]:
#bert part is not working, downside

## Summarization using bert model

In [33]:
from transformers import BertLMHeadModel, BertTokenizer

In [59]:
# Initialize the BERT model and tokenizer
bert_model = BertLMHeadModel.from_pretrained("bert-base-uncased")
bert_tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
x = bert_tokenizer(df_articles, truncation=True, max_length=500)

If you want to use `BertLMHeadModel` as a standalone, add `is_decoder=True.`


In [74]:
def bert_summarize(input_text, num_beams=4, num_words=200):
    #input_text = str(input_text)
    input_text = ' '.join(input_text.split())
    input_ids = bert_tokenizer.encode(input_text, return_tensors="pt", truncation=True)
    summary_ids = bert_model.generate(input_ids, max_length=int(num_words), min_length=80, length_penalty=2.0, num_beams=4, early_stopping=True)
    bert_summary = [bert_tokenizer.decode(g, skip_special_tokens=True,clean_up_tokenization_spaces=False ) for g in summary_ids]
    return bert_summary[0]

In [75]:
%%time
for i in range(4):
    
        print('news article ',i + 1, " : \n" )
        print(bert_summarize(df_articles[i]))
        print('............................................................................\n\n\n\n')

news article  1  : 

bishop john folda , of north dakota , is taking time off after being diagnosed . he contracted the infection through contaminated food in italy . church members in fargo , grand forks and jamestown could have been exposed . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
............................................................................




news article  2  : 



KeyboardInterrupt: 

In [85]:
# Limit the number of articles and batch them together
NUM_ARTICLES = 5
articles_batch = [article for article in df_articles[:NUM_ARTICLES]]

def bert_summarize_multiple(articles_batch, num_beams=4, num_words=200):
    input_texts = [' '.join(article.split()) for article in articles_batch]
    input_tokenized = bert_tokenizer.batch_encode_plus(input_texts, return_tensors='pt', max_length=512, truncation=True, padding='longest')

    summary_ids = bert_model.generate(input_tokenized.input_ids, max_length=int(num_words), min_length=80, length_penalty=2.0, num_beams=4, early_stopping=True)
    bert_summaries = [bert_tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=False) for g in summary_ids]
    return bert_summaries

# Generate summaries for a batch of articles
batch_summaries = bert_summarize_multiple(articles_batch)
for i, summary in enumerate(batch_summaries):
    print(f'News Article {i+1}: \n')
    print(summary)
    print('............................................................\n\n\n')


A decoder-only architecture is being used, but right-padding was detected! For correct generation results, please set `padding_side='left'` when initializing the tokenizer.


News Article 1: 

bishop john folda , of north dakota , is taking time off after being diagnosed . he contracted the infection through contaminated food in italy . church members in fargo , grand forks and jamestown could have been exposed . dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota dakota
..

In [50]:
data['article'][0]

"By . Associated Press . PUBLISHED: . 14:11 EST, 25 October 2013 . | . UPDATED: . 15:36 EST, 25 October 2013 . The bishop of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A virus in late September and early October. The state Health Department has issued an advisory of exposure for anyone who attended five churches and took communion. Bishop John Folda (pictured) of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A . State Immunization Program Manager Molly Howell says the risk is low, but officials feel it's important to alert people to the possible exposure. The diocese announced on Monday that Bishop John Folda is taking time off after being diagnosed with hepatitis A. The diocese says he contracted the infection through contaminated food while attending a conference for newly ordained 

In [51]:
data['highlights'][0]

'Bishop John Folda, of North Dakota, is taking time off after being diagnosed .\nHe contracted the infection through contaminated food in Italy .\nChurch members in Fargo, Grand Forks and Jamestown could have been exposed .'

In [73]:
def summarize_text(text):
    # Generate summary using the bart model
    summary = bert_summarize(text)
    
    # Show the generated summary
    print("\nSummary:")
    print(summary)

# Prompt the user for input and generate summary
user_input = input("Enter the text to summarize: ")
summarize_text(user_input)

Enter the text to summarize: "By . Associated Press . PUBLISHED: . 14:11 EST, 25 October 2013 . | . UPDATED: . 15:36 EST, 25 October 2013 . The bishop of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A virus in late September and early October. The state Health Department has issued an advisory of exposure for anyone who attended five churches and took communion. Bishop John Folda (pictured) of the Fargo Catholic Diocese in North Dakota has exposed potentially hundreds of church members in Fargo, Grand Forks and Jamestown to the hepatitis A . State Immunization Program Manager Molly Howell says the risk is low, but officials feel it's important to alert people to the possible exposure. The diocese announced on Monday that Bishop John Folda is taking time off after being diagnosed with hepatitis A. The diocese says he contracted the infection through contaminated food while attending a c

C:\Users\ASUS\anaconda3\envs\notebook-6.5.4\Lib\site-packages\transformers\generation\utils.py:1268: UserWarning: Input length of input_ids is 197, but `max_length` is set to 100. This can lead to unexpected behavior. You should consider increasing `max_new_tokens`.
  warnings.warn(


RuntimeError: The expanded size of the tensor (100) must match the existing size (198) at non-singleton dimension 0.  Target sizes: [100].  Tensor sizes: [198]

In [87]:
#bert part finished

<h2>evaluation metric

In [20]:
!pip install rouge-score
!pip install rouge

  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24972 sha256=7a3bfc31d5bff6436d3850f110bf7897518efc74d8836672716ba552d68b9162
  Stored in directory: c:\users\asus\appdata\local\pip\cache\wheels\1e\19\43\8a442dc83660ca25e163e1bd1f89919284ab0d0c1475475148
Successfully built rouge-score


In [76]:
from rouge import Rouge

# Define the generated summary and reference summary
generated_summary_T5 = '''The bishop of the Fargo Catholic Diocese in North Dakota has exposed church members to the hepatitis A virus in late September and early October. The risk is low, but officials feel it's important to alert people to the possible exposure.'''
reference_summary = '''Bishop John Folda, of North Dakota, is taking time off after being diagnosed .\nHe contracted the infection through contaminated food in Italy .\nChurch members in Fargo, Grand Forks and Jamestown could have been exposed .'''
# Initialize Rouge scorer
rouge_scorer = Rouge()

# Calculate ROUGE scores
rouge_scores = rouge_scorer.get_scores(generated_summary_T5,
                                       reference_summary)


print("ROUGE Scores:")
print(rouge_scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.24242424242424243, 'p': 0.22857142857142856, 'f': 0.2352941126513842}, 'rouge-2': {'r': 0.0, 'p': 0.0, 'f': 0.0}, 'rouge-l': {'r': 0.21212121212121213, 'p': 0.2, 'f': 0.20588234794550186}}]


In [80]:
from rouge import Rouge

# Define the generated summary and reference summary

generated_summary_bart = '''Fleetwood are the only team still to have a 100% record in Sky Bet League One after a 2-0 win over Scunthorpe. Peterborough, Bristol City, Chesterfield and Crawley all hit their first stumbling blocks. Posh were defeated 2-1 by Sheffield United, who had lost both of their opening contests. Bristol City were held to a go'''
reference_summary = '''Fleetwood top of League One after 2-0 win at Scunthorpe .\nPeterborough, Bristol City, Chesterfield and Crawley all drop first points of the season .\nStand-in striker Matt Done scores a hat-trick as Rochdale thrash Crewe 5-2 .\nWins for Notts County and Yeovil .\nCoventry/Bradford and Oldham/Port Vale both end in draws .\nA late Stephen Bywater own goal denies Gillingham three points against Millwall .'''
# Initialize Rouge scorer
rouge_scorer = Rouge()

# Calculate ROUGE scores
rouge_scores = rouge_scorer.get_scores(generated_summary_bart,
                                       reference_summary)


print("ROUGE Scores:")
print(rouge_scores)

ROUGE Scores:
[{'rouge-1': {'r': 0.3508771929824561, 'p': 0.4, 'f': 0.37383177072233387}, 'rouge-2': {'r': 0.16666666666666666, 'p': 0.18181818181818182, 'f': 0.1739130384877128}, 'rouge-l': {'r': 0.3508771929824561, 'p': 0.4, 'f': 0.37383177072233387}}]


## Text Summarization Project Report
### Introduction
Text summarization, a task in natural language processing, aims to create concise summaries from large texts. This report presents an approach using T5 and BART models for text summarization. We discuss their results and potential applications.

### Approach
Data Preparation:

We collected news articles and corresponding highlights, then preprocessed the data to ensure it was suitable for summarization.

### Models Used
##### T5 Model
Utilized Model: "Einmalumdiewelt/T5-Base_GNAD" variant.
Process:
T5 tokenizer used to tokenize articles and generate input tokens.
Implemented "t5_summarize" function for generating summaries with adjustable parameters.
Evaluation:
Assessed using ROUGE scores to measure summary quality against reference summaries.
##### BART Model
Model Utilized: "facebook/bart-large-cnn" variant.
Process:
BART tokenizer used to tokenize articles and create input tokens.
"bart_summarize" function implemented for generating summaries with adjustable parameters.
Evaluation:
ROUGE scores used to measure the quality of the model-generated summaries.
### Results
ROUGE Scores:

T5 Model:

ROUGE-1: Precision (0.23), Recall (0.24), F1-score (0.24).
ROUGE-2: Precision (0.0), Recall (0.0), F1-score (0.0).
ROUGE-L: Precision (0.2), Recall (0.21), F1-score (0.21).
BART Model:

ROUGE-1: Precision (0.4), Recall (0.35), F1-score (0.37).
ROUGE-2: Precision (0.18), Recall (0.17), F1-score (0.17).
ROUGE-L: Precision (0.4), Recall (0.35), F1-score (0.37).
### Discussion
The BART model outperformed the T5 model across all ROUGE metrics (ROUGE-1, ROUGE-2, and ROUGE-L) in terms of F1-score. Based on these scores, BART generated higher quality summaries compared to T5.

### Applications
News Aggregation: Provides a quick overview of current events from summarized news articles.
Document Summarization: Assists in identifying key information from lengthy documents or research papers.
Social Media Analysis: Aids in understanding trends and sentiment analysis from social media content.
Content Generation: Useful in producing concise summaries as part of content generation systems.
### Conclusion
In conclusion, the BART model consistently outperformed the T5 model in the text summarization task. With higher F1-scores across all key metrics, BART demonstrates superior performance in generating high-quality summaries.